Show-LastPassConsoleGridView + Get-VaultParams changes

itfranck opened the pull request:
A "candy" cmdlet that integrate Microsoft.PowerShell.ConsoleGuiTools. 
You might know of this module ? If not, see [this](https://devblogs.microsoft.com/powershell/introducing-consoleguitools-preview/) 😉 

I think that will be probably one of my favorite way to go about consulting the vault from now on. 
I didn't want to add the module as a required dependency so instead, an error is thrown if it is used without the module installed.
A warning is also thrown if you are using 5.1  or anything below 6.2 for that matter.

All that would be missing is a cute Gif with it in action (using a test vault). 
You can create a one line PS script and use it whenever with the KeepOpen parameter and / or use it interactively to select a secret and return it unformatted with PassThru.

I think this tie up everything nicely as you now have a one-liner interface at your fingertips.

Edit: Oh yeah, I changed `Get-VaultParams` for `Get-SelectedVault` which does essentially the same thing but return the vault info instead of the parameter since I wanted the error management from that cmdlet into the new cmdlet but I also needed to get back the name of the vault and not its parameters.

itfranck committed:
Show-LastPassConsoleGridView + Get-VaultParams changes

itfranck committed:
Extra line removal

TylerLeonhardt commented:
Love this idea :) should we maybe allow folks to use Out-GridView as a fall back if they don't have ocgv?

itfranck committed:
Out-GridView fallback

itfranck commented:
@TylerLeonhardt Done. I love that.

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -216,6 +216,85 @@ function Get-VaultParams {
+        $UseConsoleGridView = $true
+    }
+    catch {
+        Write-Debug "Microsoft.Powershell.ConsoleGuiTools could not be loaded.`n$($_ | Out-String)"
```
Inside of this catch, you still need to check if `Out-GridView` is available because on macOS & Linux that cmdlet doesn't exist. Perhaps even saying `Can't find a grid view cmdlet. Try installing the 'Microsoft.PowerShell.ConsoleGuiTools' module and try again.`

itfranck replied:
https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/23/commits/c30cc6f8fee4c660b96452aeadd7d32cfd9114ea

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -216,6 +216,85 @@ function Get-VaultParams {
+This cmdlet can make use of the improved Out-ConsoleGridView cmdlet if using Powershell 6.2 or newer and  Microsoft.PowerShell.ConsoleGuiTools is installed.
+Otherwise, Out-GridView will be used.
+#>
+Function Show-LastPassConsoleGridView {
```
```suggestion
function Show-LastPassConsoleGridView {
```

I think we have lowercase in this file

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -216,6 +216,85 @@ function Get-VaultParams {
+    [CmdletBinding(DefaultParameterSetName='Default')]
+    Param(
+        [String]$Vault,
+        [String]$Filter = '*',
```
You add star to the end of $Filter down below so I don't think you need to set this default for this parameter since:

```
"$filter" -eq ""
```
when `$Filter` is undefined

itfranck replied:
https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/23/commits/a601d6098f5600db1249874a5564b8ed965c5716

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -216,6 +216,85 @@ function Get-VaultParams {
+        [Parameter(ParameterSetName = 'KeepOpen')]
+        [Switch]$KeepOpen,
+        [Parameter(ParameterSetName = 'PassThru')]
+        [Switch]$PassThru
```
Why do you need this `PassThru`? I think it should be the default behavior to emit the result of the choice to the pipeline.

itfranck replied:
Yeah, I can do that. 
My beef was / is with the multiline notes value. 
https://github.com/TylerLeonhardt/SecretManagement.LastPass/issues/24

I can't figure out (yet) how to apply a view formatter for a Idictionary 
(This is the only topic where I find there's a lack of examples to do advanced stuff with)

Ideally, I'd have a view formatter in place to auto-expand that property.
I can remove PassThru and instead replace it with a parameter to do the opposite.

itfranck replied:
@TylerLeonhardt What do you think if I do the exact opposite of that commit. 
Meaning `-Passthru` become `-Formatted`. 
By default (without switch), secret is returned as is, perfect for variable assignement.

`Formatted` will include the extra treatment, which essentially make sure that the `Notes` field is expanded and readable from the output. 


```
 # By default, return the secret as is, everything will be fine
            if (!$Formatted) {return $Secret}

            # Intended for view, not for assignement. Add secret title and expand multiline notes in the console.
                Write-host $_.Name -ForegroundColor Cyan
                # Expand multiline notes (Ideally replaced by a format view at some point)
                if ($null -ne $Secret.Notes -and $Secret.Notes.IndexOf("`n") -ne -1) {
                    $Notes = $Secret.Notes
                    $Secret.Remove('Notes')
                    $Secret
                    "Notes:`n$Notes"
                }
                else {
                    $Secret
                }
```

Essentially, this provide the kind of readability that using a view formatter on a collection would do (#24) except that I am not sure how to do a view formatter on a collection. 

I would be able to manage on a PSObject but since we have case-sensitivity, I need to stick to ordered dictionaries. 

If I could have a type or view formatter that expand the Notes field, then all that Formatted bit would be superfluous.

itfranck replied:
tl;dr : I'd remove the PassThru, create a Formatted switch so that it reverse the default output.
Default output would be raw secret. 
Formatted output would ensure Notes field is expanded so you can read it from the screen.
All of that because I can't come with a way (yet) to solve #24

TylerLeonhardt replied:
> I would be able to manage on a PSObject but since we have case-sensitivity, I need to stick to ordered dictionaries.

You can cast an OrderedDictionary to a PSObject. `[PSObject]$dict` or `[pscustomobject]$dict`. Technically all objects in PowerShell are secretly a PSObject.

itfranck replied:
Hmmm ... Good point. 
I'll check if I can do what I want from that. 

If I can have a LastPassSecret type inheriting from ordered dictionary and set a view to it so it return the Notes expanded by default without the need of "Format output altering" command, then it would be golden.

TylerLeonhardt replied:
Something they did in [PowerShell for GitHub](https://github.com/Microsoft/PowerShellForGitHub) is [have their Format.ps1xmls as normal](https://github.com/microsoft/PowerShellForGitHub/blob/master/Formatters/GitHubGists.Format.ps1xml) and then tack on the type name like so:

```
$item.PSObject.TypeNames.Insert(0, $TypeName)
```
https://github.com/microsoft/PowerShellForGitHub/blob/9600fc21120e17241e60606c5de3459d973026bb/GitHubGists.ps1#L1750

itfranck replied:
Ah Thanks. 
I was playing around this concept just now. 

Your "Everything is a PSObject" quote put me on track.
I knew that but it was buried somewhere. 

```
$test  = [System.Collections.Specialized.OrderedDictionary]::new([System.StringComparer]::CurrentCulture)
$test.Add('ss','aa')
$test.Add('aaa','aa')
$test.Add('aaA', 'aa')
$test.PSObject.TypeNames.Insert(0,'LastPassSecret')
```

I am trying to flesh out some kind of ps1xml that would work and playing around with EzOut helper.


```
Write-FormatView -TypeName 'LastPassSecret' -Property 'Name','Value' -Width 1,1 | Out-FormatData | Add-FormatData; $test
```

```
<Configuration>
  <ViewDefinitions>
    <View>
      <Name>LastPassSecret</Name>
      <ViewSelectedBy>
        <TypeName>LastPassSecret</TypeName>
      </ViewSelectedBy>
      <TableControl>
        <TableHeaders>
          <TableColumnHeader>
            <Alignment>left</Alignment>
            <Width>1</Width>
          </TableColumnHeader>
          <TableColumnHeader>
            <Alignment>left</Alignment>
            <Width>1</Width>
          </TableColumnHeader>
        </TableHeaders>
        <TableRowEntries>
          <TableRowEntry>
            <TableColumnItems>
              <TableColumnItem>
                <PropertyName>Name</PropertyName>
              </TableColumnItem>
              <TableColumnItem>
                <PropertyName>Value</PropertyName>
              </TableColumnItem>
            </TableColumnItems>
          </TableRowEntry>
        </TableRowEntries>
      </TableControl>
    </View>
  </ViewDefinitions>
</Configuration>
```

It's not working yet though.
My thought is that if I can mess up the width of the column, I'll have succeeded in getting the proper tree syntax and I'll then dig around the Scriptblock property to get this working.

Maybe I'm owed a visit to the format.xsd schema to see if I can gain more insight in the proper manner to write something that would affect the collection.

itfranck replied:
I was unsuccessful with the view formatter. 
This is a concept I'm slowly wrapping my head around but I find there's a lack of content compared to anything else on the matter so that's something I'll look into implementing at a latter time, if I can figure out the proper way to write the formatter. 

https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/23/commits/f695769b3ea13a3a53cadef962c1645b1bcf1158

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -216,6 +216,85 @@ function Get-VaultParams {
+        if ($null -eq $Result) { break }
+        $Result | ForEach-Object { 
+            $Secret = Microsoft.Powershell.SecretManagement\Get-Secret -Vault $Vault -Name $_.Name -AsPlainText 
+            if ($PassThru) {return $Secret}
```
do you need this if it gets emitted below?

itfranck replied:
Not sure if still relevant with the last commit.
https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/23/commits/f695769b3ea13a3a53cadef962c1645b1bcf1158

The Formatted switch is there to provide a readable expanded output. 
Default output is now what the "Passthru" initially did.

itfranck committed:
Additional error check for Grid view.

itfranck committed:
Fix error check

itfranck committed:
Moved check inside Catch

itfranck committed:
Syntax

Co-authored-by: Tyler James Leonhardt <tylerl0706@gmail.com>

itfranck committed:
Show-LastPassConsoleGridView - Filter default value

TylerLeonhardt commented on SecretManagement.LastPass.psm1:
```diff
@@ -216,6 +216,88 @@ function Get-VaultParams {
+        }
+        
+        if ($null -eq $Result) { break }
+        $Result | ForEach-Object { 
```
`$result` should be only one item since you specified `-OutputMode Single` above

itfranck replied:
https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/23/commits/f695769b3ea13a3a53cadef962c1645b1bcf1158

itfranck committed:
Formatted instead of PassThru

itfranck committed:
Parameter set name not required anymore.

itfranck committed:
Default outputType PSCredentials management for formatted view

itfranck commented:
@TylerLeonhardt I think I addressed the remaining points. 
I also added a formatted output if we get a pscredential result (It will happens with the default outputType, which I had not considered) 

https://github.com/TylerLeonhardt/SecretManagement.LastPass/pull/23/commits/46ed910c1e7c41d5e1cbf65d0c00ae17d54fa1da

itfranck committed:
Show-LastPassGridView

itfranck commented:
One last thing, I changed the name from `Show-LastPassConsoleGridView` to `Show-LastPassGridView`
- We won't necessarily use the ConsoleGridView  (fallback)
- ConsoleGridView is still a GridView
- Show-LastPassGridView is shorter.

TylerLeonhardt reviewed (approved):
LGTM! Merge the 2 PRs in and I'll get going on a release tonight/tomorrow

TylerLeonhardt commented:
Released! https://twitter.com/TylerLeonhardt/status/1330335754791239685?s=20

TylerLeonhardt commented:
btw, you may notice that I added a devcontainer to this repo so you can either use the Remote-Container extension for vscode locally, or use GitHub Codespaces. Both will make sure all the right deps are installed. I was able to start the container, login with lpass (which was already installed) and run the tests and it all worked.
